### PR TITLE
New Adapter: Limelight Digital

### PR DESCRIPTION
**Pull request by imedvedko**
Docs: https://github.com/prebid/prebid.github.io/pull/4297

### CODE REVIEW

**Comment by bretg** [comment]
As noted in the docs PR https://github.com/prebid/prebid.github.io/pull/4297, please update this adapter to be able to support the same set of parameters as your client-side adapter.

**Comment by imedvedko** [comment]
ping :-)
As noted in the docs PR https://github.com/prebid/prebid.github.io/pull/4297, we updated this adapter to be able to support the same set of parameters as your client-side adapter. Could you please review it?

File: static/bidder-info/iionads.yaml
Comment (VeronikaSolovei9) [inline static/bidder-info/iionads.yaml:1]: This is so weird to see URL like this, but I guess it's possible too. 
Do you want to add validation for the "host" parameter to make sure it's URL, not just a random set of characters?

Comment (VeronikaSolovei9) [reply static/bidder-info/iionads.yaml:1]: Thank you. We discussed this with the team. Unfortunately this URL format may lead to a potential security issue. We recommend to modify it to something like this (based on unit tests you provided):
"https://www.limelightDigital.{{.MediaType}}.test.com/openrtb/{{.PublisherID}}"

Comment (SyntaxNode) [reply static/bidder-info/iionads.yaml:1]: Correct @VeronikaSolovei9. We do not allow the domain name to be configurable due to security issues. Additionally, we highly discourage subdomains as they negatively affect the performance of maintaining connects to your bidding server. Please see our bidder development guide for further information: https://docs.prebid.org/prebid-server/developers/add-new-bidder-go.html

I am aware there are a few adapters currently in violation. The Prebid Server Committee is working on a remediation plan.

Comment (imedvedko) [reply static/bidder-info/iionads.yaml:1]: As noted in the docs PR https://github.com/prebid/prebid.github.io/pull/4297, we need to support the same parameters with our Prebid.js adapter. Our Prebid.js has `Host` parameter. And a lot of our client use it on their custom domains which are owned by our clients and which we do not directly control. Are you suggesting we set up cname dns records from our global domain to their custom domains and hope to win performance?

Comment (imedvedko) [reply static/bidder-info/iionads.yaml:1]: Is it perfect solution if we will create a new alias adapter for each our client with his own domain? No subdomains. Ho `Host` parameters. Will it be easier to maintain? Will you win in performance and connections?

File: adapters/limelightDigital/limelightDigital.go
Comment (Sonali-More-Xandr) [inline adapters/limelightDigital/limelightDigital.go:131]: Helper function need not be defined on an adapter. It can simply defined as:-

`func getImpressionExt(imp *openrtb2.Imp) (*openrtb_ext.ImpExtLimelightDigital, error)`

File: adapters/limelightDigital/limelightDigital.go
Comment (Sonali-More-Xandr) [inline adapters/limelightDigital/limelightDigital.go:168]: Could you please help me understand why the BidTypeBanner is returned in this case as well on line 171?

File: adapters/limelightDigital/limelightDigital.go
Comment (Sonali-More-Xandr) [inline adapters/limelightDigital/limelightDigital.go:148]: This function need not be defined on adapter. This can be done as follow:-

`func buildEndpointURL(params *openrtb_ext.ImpExtLimelightDigital) (string, error)`

File: adapters/limelightDigital/LimelightDigitaltest/supplemental/empty_object reponse.json
Comment (Sonali-More-Xandr) [inline adapters/limelightDigital/LimelightDigitaltest/supplemental/empty_object reponse.json:1]: Nitpick - missing underscore ("_") in the filename. Recommend renaming it as :-
empty_object_response.json

Comment (imedvedko) [reply adapters/limelightDigital/LimelightDigitaltest/supplemental/empty_object reponse.json]: Sorry. I fixed it in https://github.com/prebid/prebid-server/pull/2539/commits/a62aea11c7a63b52f404abd5cd74592861825fb3 commit

File: adapters/limelightDigital/limelightDigital.go
Comment (imedvedko) [reply adapters/limelightDigital/limelightDigital.go]: I used `endpointTemplate` adapter param on 153 line. I rewrite it and receive template as argument in https://github.com/prebid/prebid-server/pull/2539/commits/b3d20782a84aaa4c9c184bbfb0dd9c896d56e7b9 commit. Thanks

File: adapters/limelightDigital/limelightDigital.go
Comment (imedvedko) [reply adapters/limelightDigital/limelightDigital.go]: Thank you so much for your review It really looks like an opportunity for improvement. I improve BidType checking in https://github.com/prebid/prebid-server/pull/2539/commits/a8efcf771fdf43622499d870cd6f88a7e92c9bf2 and create a test which checks this case

Comment (imedvedko) [reply adapters/limelightDigital/limelightDigital.go]: I renamed test in https://github.com/prebid/prebid-server/pull/2539/commits/fb67dbd204813d43f18ae5a662b523d1e8731fd3 commit

File: adapters/limelightDigital/limelightDigital.go
Comment (VeronikaSolovei9) [reply adapters/limelightDigital/limelightDigital.go]: I'm not sure about this change. For other adapters  `buildEndpointURL` gets an adapter as a receiver. Please refer to usages of this function in other adapters.

File: adapters/limelightDigital/limelightDigital.go
Comment (VeronikaSolovei9) [reply adapters/limelightDigital/limelightDigital.go]: Good point. With new changes L174 can be fully removed or modified to `break`
Also please add test for `native` media type

File: adapters/limelightDigital/limelightDigital.go
Comment (VeronikaSolovei9) [inline adapters/limelightDigital/limelightDigital.go:56]: please add test for this case

File: adapters/limelightDigital/limelightDigital.go
Comment (imedvedko) [reply adapters/limelightDigital/limelightDigital.go:154]: @VeronikaSolovei9 Thank you. I revert these changes in https://github.com/prebid/prebid-server/pull/2539/commits/4af1869e6e0d25e15621ec663aa6241e25b43cda commit. Thanks

File: adapters/limelightDigital/limelightDigital.go
Comment (imedvedko) [reply adapters/limelightDigital/limelightDigital.go]: @VeronikaSolovei9 I improved unknown media type filtering in commit  https://github.com/prebid/prebid-server/pull/2539/commits/0564ec79541d9b339727fcfe1f29aceb5e150698 commit and added native media type test in https://github.com/prebid/prebid-server/pull/2539/commits/25733951529ac9f1415931d326d3955a93978729 commit

File: adapters/limelightDigital/limelightDigital.go
Comment (imedvedko) [reply adapters/limelightDigital/limelightDigital.go:56]: Thanks for your comments. I added the test in https://github.com/prebid/prebid-server/pull/2539/commits/ef1e12ee84b4deedffd9d75323bf497629177837 commit. Could you check it and resolve the conversation?

File: openrtb_ext/imp_limelightDigital.go
Comment (Sonali-More-Xandr) [inline openrtb_ext/imp_limelightDigital.go:5]: The [doc PR](https://github.com/prebid/prebid.github.io/pull/4297/files) has required fields adUnitId and adUnitType which are missing here. If those fields are required, could you please consider adding those here as well, if not, then please update the doc PR to match the parameters.

File: static/bidder-info/iionads.yaml
Comment (Sonali-More-Xandr) [inline static/bidder-info/iionads.yaml:6]: The mediaTypes mentioned in the [doc PR](https://github.com/prebid/prebid.github.io/pull/4297/files) are not matching. Could you please update the doc PR to reflect all the supported media types.

File: openrtb_ext/imp_limelightDigital.go
Comment (imedvedko) [reply openrtb_ext/imp_limelightDigital.go:5]: @Sonali-More-Xandr It is required fields for our Prebid.js dapter. As noted in the docs PR https://github.com/prebid/prebid.github.io/pull/4297, we need to support the same set of parameters with our Prebid.js adapter.

Comment (imedvedko) [reply openrtb_ext/imp_limelightDigital.go:5]: We can not remove these fields from doc because are required fields for our Prebid.js adapter. And we don't need to use these filed in our Prebid Server adapter. Should we add parameters that we don't use to our Prebid Server adapter?

Comment (bretg) [reply openrtb_ext/imp_limelightDigital.go:5]: The server should be able to accept the same parameters, ignoring them if not needed server-side.

This is to allow publishers to use the server-to-server test module, which uses the same set of configuration for your bidder.

File: static/bidder-info/iionads.yaml
Comment (imedvedko) [reply static/bidder-info/iionads.yaml:6]: As noted in the docs PR https://github.com/prebid/prebid.github.io/pull/4297, we have to use the same documentation with our Prebid.js adapter. This means that we must specify in the documentation only the set of media types that both our Prebid.js and Prebid Server adapters support.

Comment (bretg) [reply static/bidder-info/iionads.yaml:6]: It would be fine to add a note to your docs to explain this. e.g. "the client-side adapter doesn't support X, but the server-side adapter does."

File: openrtb_ext/imp_limelightDigital.go
Comment (imedvedko) [reply openrtb_ext/imp_limelightDigital.go:5]: @bretg Thanks, you are absolutely right. The server is already able to accept the same parameters, it ignore them. I added tests with this case in https://github.com/prebid/prebid-server/pull/2539/commits/0b780e88c6cda75ce2fa382633a7959061224e4b commit

File: static/bidder-info/iionads.yaml
Comment (imedvedko) [reply static/bidder-info/iionads.yaml:6]: @bretg Thank you for your comment. I added this note in https://github.com/prebid/prebid.github.io/pull/4297/commits/d7223cf3e374f2e9a4e085283b63fcfe944c34a8 commit

File: static/bidder-info/iionads.yaml
Comment (bsardo) [reply static/bidder-info/iionads.yaml:1]: Hi @imedvedko. Thanks for your patience and for working with us on this feedback.

The two main concerns we have for using a macro in the domain are security exposure and connection performance. The security aspect is alleviated by using a fixed top level domain. Due to the potential harm to hosts, we are strict in this requirement. We are working towards fixing  the few adapters currently in violation.

The connection performance advice is for your benefit. Client specific subdomains prevent Prebid Server from reusing connections across your clients which results in more connections needed for your adapter. The issue gets worse the more successful you are in attracting new clients. Hosts may choose to disable your adapter due to this behavior.

The best practice is for your domain to be fixed and any parameters specified in the path or query. How you choose to forward bid requests to your clients is outside of the scope of PBS and for your team to decide. I am hesitant to advise on this but my gut is that cnames are not the answer but perhaps it could be accomplished with vhosts, url rewrites or some other means at the edge.

Please do not create an alias for each client. That is not the intention of this feature.

I hope that helps.

Comment (hhhjort) [reply static/bidder-info/iionads.yaml:1]: Also note that only the PBS host can see the URL used to connect to your bidder. Well, a debug auction can reveal the http call(s) sent out to the bidder, but beyond that it is completely invisible. So there is little vanity reason for having multiple URLs. A structural reason may exist in the event that each URL goes to a different physical host which each have a different dataset loaded to represent each client. But as Brian points out above, this doesn't seem like a very scalable architecture.

Comment (imedvedko) [reply static/bidder-info/iionads.yaml]: Hi, sorry for the long answer. I added top level domain. For compatibility with our Prebid.js adapter parameters, I left `host` field format in the request the same and just cut out subdomain from it. I also added additional validation of `host` parameter to improve the security of subdomain cutting out https://github.com/prebid/prebid-server/pull/2539/commits/e0b1ab238714f95dea2150f81d3ca3ab4f65b34b#diff-85e257afc296b1b9772d361fe8269082ee7a158207ed84f6fae16e1b42a2426cR12 and added additional params tests https://github.com/prebid/prebid-server/pull/2539/commits/e0b1ab238714f95dea2150f81d3ca3ab4f65b34b#diff-98d261a63ffeb65e201644a1669caaa0d9e6c2fb2a9a9e6de7d429d22814b017R61-R70 . Thanks

Comment (imedvedko) [reply static/bidder-info/iionads.yaml]: I have improved parameters validation through json schema in previous commits. I also added additional domain validation for host parameter in the adapter code in https://github.com/prebid/prebid-server/pull/2539/commits/7b2cc3abd0eb757108d22990267240fa63ca9351 commit. Feel free to request me to rollback this commit if these changes seem redundant to you

Comment (hhhjort) [reply static/bidder-info/iionads.yaml]: This should prevent PBS from being used as a malicious proxy, unless a PBS host or future PR does something silly like define the endpoint as "https://{{.HOST}}.com/openrtb/..." It still has the potential of cluttering up the connection pool depending on how many unique HOST values the adapter might see as it runs on a host. But at least that is just a performance issue rather than a security issue.

**Comment by bsardo** [comment]
Hi @imedvedko, the hold up has been about the macros in your endpoint. With the adjustments you recently made you are within the rules for now, however things will be a bit more strict later this year. I invite you to participate in the discussion in [Prebid Server Rules Enforcement #2437: Dynamic Host Names](https://github.com/prebid/prebid-server/issues/2606) and to help us better understand your use case.

The team will pick this PR up again next week for review. Thanks for your patience.

**Review by Sonali-More-Xandr — APPROVED** [review]
Good to go after one more approval.

File: static/bidder-info/iionads.yaml
Comment (onkarvhanumante) [inline static/bidder-info/iionads.yaml:1]: should this be `https` ? similar thoughts for `static/bidder-info/limelightDigital.yaml` endpoint

Comment (imedvedko) [reply static/bidder-info/iionads.yaml:1]: As I can see, it is just not recommended, but it's not prohibited. As I see, a lot of Prebid Server adapters do the same.

Comment (imedvedko) [reply static/bidder-info/iionads.yaml:1]: Is it blocking the pull request merge?

Comment (onkarvhanumante) [reply static/bidder-info/iionads.yaml:1]: Although not prohibited now.  But if possible should prefer using `https` over `http`.  Let us know this change is possible. If not then we can merge PR change.

Comment (imedvedko) [reply static/bidder-info/iionads.yaml:1]: We use `http` for server-to-server communication. Please merge the pull request as it is, if it is possible :-)

Comment (bretg) [reply static/bidder-info/iionads.yaml:1]: HTTP is tolerated for server-to-server. It's understood that it's more expensive to run load balancers that process SSL.
Client-side adapters must use HTTPS.

Comment (imedvedko) [reply static/bidder-info/iionads.yaml:1]: @bretg You are absolutely right, so we use `https` in our client side Prebid.js adaper https://github.com/prebid/Prebid.js/blob/master/modules/limelightDigitalBidAdapter.js#L119

**Review by onkarvhanumante — APPROVED** [review]

**Comment by imedvedko** [comment]
Thank you so much for the time of each of you and for the deep desire to dive into our needs. :-)

**Pull request merged**
